src/Datadog.Trace/ScopeManager.cs

```diff
@@ -0,0 +1,143 @@
+                    for (int i = 0; i < _prioritizedScopeAccess.Count; i++)
+                    {
+                        try
+                        {
+                            var activeScope = _prioritizedScopeAccess[i].GetActiveScope();
+                            if (activeScope != null)
+                            {
+                                return activeScope;
+                            }
+                        }
+                        catch (Exception ex)
+                        {
+                            Log.ErrorException($"Active scope access failed: {_prioritizedScopeAccess[i]?.GetType().FullName}.", ex);
+                            continue;
+                        }
+                    }
+                }
+
+                return null;
+            }
+        }
+
+        public Scope Activate(Span span, bool finishOnClose)
+        {
+            var newParent = Active;
+            var scope = new Scope(newParent, span, this, finishOnClose);
+            var scopeOpenedArgs = new SpanEventArgs(span);
+
+            SpanOpened?.Invoke(this, scopeOpenedArgs);
+
+            SetActiveScope(scope);
+
+            if (newParent != null)
+            {
+                SpanDeactivated?.Invoke(this, new SpanEventArgs(newParent.Span));
+            }
+
+            SpanActivated?.Invoke(this, scopeOpenedArgs);
+
+            return scope;
+        }
+
+        public void Close(Scope scope)
+        {
+            var current = Active;
+            var isRootSpan = scope.Parent == null;
+
+            if (current == null || current != scope)
+            {
+                // This is not the current scope for this context, bail out
+                return;
+            }
+
+            // if the scope that was just closed was the active scope,
+            // set its parent as the new active scope
+            SetActiveScope(current.Parent);
+            SpanDeactivated?.Invoke(this, new SpanEventArgs(current.Span));
+
+            if (!isRootSpan)
+            {
+                SpanActivated?.Invoke(this, new SpanEventArgs(current.Parent.Span));
+            }
+
+            SpanClosed?.Invoke(this, new SpanEventArgs(scope.Span));
+
+            if (isRootSpan)
+            {
+                TraceEnded?.Invoke(this, new SpanEventArgs(scope.Span));
+            }
+        }
+
+        public void RegisterScopeAccess(IActiveScopeAccess scopeAccess)
+        {
+            lock (_scopeAccessLock)
+            {
+                if (_prioritizedScopeAccess.Any(i => i.GetType() == scopeAccess.GetType()))
+                {
+                    // We don't want multiple instances as each type manages scope in an ambient fashion.
+                    return;
+                }
+
+                _prioritizedScopeAccess.Add(scopeAccess);
+
+                // Slower add for faster read
+                _prioritizedScopeAccess = _prioritizedScopeAccess.OrderByDescending(i => i.Priority).ToList();
```

Comment:
Since `_prioritizedScopeAccess` is a concrete in-memory list, we can use `List<T>.Sort()` to sort the items in place instead of using LINQ (which creates a new list). Something like:
```csharp
_prioritizedScopeAccess.Sort((x, y) => x.Priority.CompareTo(y.Priority));
```

Or if we really wanted to go for performance over readability, instead of sorting the entire list each time we add a new item, we can find the index where the new item should be inserted to keep the list sorted:
```csharp
var priorityComparer = Comparer<IActiveScopeAccess>.Create((x, y) => x.Priority.CompareTo(y.Priority));

var index = list.BinarySearch(scopeAccess, priorityComparer);
list.Insert(~index, scopeAccess);
```

Reply:
Or use `SortedSet<T>`. 🤦‍♂